File: CS/NonPersistentObjectsDemo.Module/BusinessObjects/CollectionComplete/Project.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Xml.Serialization;
using DevExpress.Data.Filtering;
using DevExpress.ExpressApp;
using DevExpress.Persistent.Base;
using DevExpress.Persistent.BaseImpl;
using DevExpress.Xpo;

namespace NonPersistentObjectsDemo.Module.BusinessObjects {

    [DevExpress.ExpressApp.DC.XafDefaultProperty(nameof(CodeName))]
    [DefaultClassOptions]
    public class Project : BaseObject, IObjectSpaceLink {
        public Project(Session session) : base(session) { }

        private string _CodeName;
        public string CodeName {
            get { return _CodeName; }
            set { SetPropertyValue<string>(nameof(CodeName), ref _CodeName, value); }
        }

        #region Features
        private BindingList<Feature> _Features;
        [Aggregated]
        public BindingList<Feature> Features {
            get {
                if(_Features == null) {
                    _Features = new BindingList<Feature>();
                    _Features.ListChanged += _Features_ListChanged;
                }
                return _Features;
            }
        }
        private void _Features_ListChanged(object sender, ListChangedEventArgs e) {
            var list = (BindingList<Feature>)sender;
            if(e.ListChangedType == ListChangedType.ItemAdded) {
                var obj = list[e.NewIndex];
                obj.OwnerKey = this.Oid;
                obj.LocalKey = e.NewIndex + 1;
            }
            FeatureList = SerializationHelper.Save(Features);
        }
        private string _FeatureList;
        [Browsable(false)]
        [Size(SizeAttribute.Unlimited)]
        public string FeatureList {
            get { return _FeatureList; }
            set { SetPropertyValue<string>(nameof(FeatureList), ref _FeatureList, value); }
        }
        #endregion

        #region MainFeature
        private Feature _MainFeature;
        [DataSourceProperty(nameof(Features))]
        public Feature MainFeature {
            get { return _MainFeature; }
            set { SetPropertyValue<Feature>(nameof(MainFeature), ref _MainFeature, value); }
        }
        private string _MainFeatureName;
        [Browsable(false)]
        public string MainFeatureName {
            get { return _MainFeatureName; }
            set { SetPropertyValue<string>(nameof(MainFeatureName), ref _MainFeatureName, value); }
        }
        #endregion

        #region Resources
        private BindingList<Resource> _Resources;
        [Aggregated]
        public BindingList<Resource> Resources {
            get {
                if(_Resources == null) {
                    _Resources = new BindingList<Resource>();
                    _Resources.ListChanged += _Resources_ListChanged;
                }
                return _Resources;
            }
        }
        private void _Resources_ListChanged(object sender, ListChangedEventArgs e) {
            var list = (BindingList<Resource>)sender;
            if(e.ListChangedType == ListChangedType.ItemAdded) {
                list[e.NewIndex].OwnerKey = this.Oid;
            }
            ResourceList = SerializationHelper.Save(Resources);
        }
        private string _ResourceList;
        [Browsable(false)]
        [Size(SizeAttribute.Unlimited)]
        public string ResourceList {
            get { return _ResourceList; }
            set { SetPropertyValue<string>(nameof(ResourceList), ref _ResourceList, value); }
        }
        #endregion

        protected override void OnChanged(string propertyName, object oldValue, object newValue) {
            base.OnChanged(propertyName, oldValue, newValue);
            if(propertyName == nameof(MainFeature)) {
                MainFeatureName = (newValue as Feature)?.Name;
            }
        }
        protected override void OnLoaded() {
            base.OnLoaded();
            int counter = 0;
            SerializationHelper.Load(Features, FeatureList, ObjectSpace, o => { o.OwnerKey = this.Oid; o.LocalKey = ++counter; });
            SerializationHelper.Load(Resources, ResourceList, ObjectSpace, o => { o.OwnerKey = this.Oid; });
            _MainFeature = MainFeatureName == null ? null : Features.FirstOrDefault(f => f.Name == MainFeatureName);
        }
        protected override void OnSaving() {
            FeatureList = SerializationHelper.Save(Features);
            ResourceList = SerializationHelper.Save(Resources);
            base.OnSaving();
        }
        private IObjectSpace objectSpace;
        protected IObjectSpace ObjectSpace { get { return objectSpace; } }
        IObjectSpace IObjectSpaceLink.ObjectSpace {
            get { return objectSpace; }
            set {
                if(objectSpace != value) {
                    objectSpace = value;
                }
            }
        }
    }
}

File: VB/NonPersistentObjectsDemo.Module/BusinessObjects/CollectionComplete/Project.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.IO
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks
Imports System.Xml.Serialization
Imports DevExpress.Data.Filtering
Imports DevExpress.ExpressApp
Imports DevExpress.Persistent.Base
Imports DevExpress.Persistent.BaseImpl
Imports DevExpress.Xpo

Namespace NonPersistentObjectsDemo.Module.BusinessObjects

	<DevExpress.ExpressApp.DC.XafDefaultProperty(NameOf(Project.CodeName))>
	<DefaultClassOptions>
	Public Class Project
		Inherits BaseObject
		Implements IObjectSpaceLink

		Public Sub New(ByVal session As Session)
			MyBase.New(session)
		End Sub

		Private _CodeName As String
		Public Property CodeName() As String
			Get
				Return _CodeName
			End Get
			Set(ByVal value As String)
				SetPropertyValue(Of String)(NameOf(CodeName), _CodeName, value)
			End Set
		End Property

		#Region "Features"
		Private _Features As BindingList(Of Feature)
		<Aggregated>
		Public ReadOnly Property Features() As BindingList(Of Feature)
			Get
				If _Features Is Nothing Then
					_Features = New BindingList(Of Feature)()
					AddHandler _Features.ListChanged, AddressOf _Features_ListChanged
				End If
				Return _Features
			End Get
		End Property
		Private Sub _Features_ListChanged(ByVal sender As Object, ByVal e As ListChangedEventArgs)
			Dim list = DirectCast(sender, BindingList(Of Feature))
			If e.ListChangedType = ListChangedType.ItemAdded Then
				Dim obj = list(e.NewIndex)
				obj.OwnerKey = Me.Oid
				obj.LocalKey = e.NewIndex + 1
			End If
			FeatureList = SerializationHelper.Save(Features)
		End Sub
		Private _FeatureList As String
		<Browsable(False)>
		<Size(SizeAttribute.Unlimited)>
		Public Property FeatureList() As String
			Get
				Return _FeatureList
			End Get
			Set(ByVal value As String)
				SetPropertyValue(Of String)(NameOf(FeatureList), _FeatureList, value)
			End Set
		End Property
		#End Region

		#Region "MainFeature"
		Private _MainFeature As Feature
		<DataSourceProperty(NameOf(Features))>
		Public Property MainFeature() As Feature
			Get
				Return _MainFeature
			End Get
			Set(ByVal value As Feature)
				SetPropertyValue(Of Feature)(NameOf(MainFeature), _MainFeature, value)
			End Set
		End Property
		Private _MainFeatureName As String
		<Browsable(False)>
		Public Property MainFeatureName() As String
			Get
				Return _MainFeatureName
			End Get
			Set(ByVal value As String)
				SetPropertyValue(Of String)(NameOf(MainFeatureName), _MainFeatureName, value)
			End Set
		End Property
		#End Region

		#Region "Resources"
		Private _Resources As BindingList(Of Resource)
		<Aggregated>
		Public ReadOnly Property Resources() As BindingList(Of Resource)
			Get
				If _Resources Is Nothing Then
					_Resources = New BindingList(Of Resource)()
					AddHandler _Resources.ListChanged, AddressOf _Resources_ListChanged
				End If
				Return _Resources
			End Get
		End Property
		Private Sub _Resources_ListChanged(ByVal sender As Object, ByVal e As ListChangedEventArgs)
			Dim list = DirectCast(sender, BindingList(Of Resource))
			If e.ListChangedType = ListChangedType.ItemAdded Then
				list(e.NewIndex).OwnerKey = Me.Oid
			End If
			ResourceList = SerializationHelper.Save(Resources)
		End Sub
		Private _ResourceList As String
		<Browsable(False)>
		<Size(SizeAttribute.Unlimited)>
		Public Property ResourceList() As String
			Get
				Return _ResourceList
			End Get
			Set(ByVal value As String)
				SetPropertyValue(Of String)(NameOf(ResourceList), _ResourceList, value)
			End Set
		End Property
		#End Region

		Protected Overrides Sub OnChanged(ByVal propertyName As String, ByVal oldValue As Object, ByVal newValue As Object)
			MyBase.OnChanged(propertyName, oldValue, newValue)
			If propertyName = NameOf(MainFeature) Then
				MainFeatureName = TryCast(newValue, Feature)?.Name
			End If
		End Sub
		Protected Overrides Sub OnLoaded()
			MyBase.OnLoaded()
			Dim counter As Integer = 0
			SerializationHelper.Load(Features, FeatureList, ObjectSpace, Sub(o)
				o.OwnerKey = Me.Oid
				counter += 1
				o.LocalKey = counter
			End Sub)
			SerializationHelper.Load(Resources, ResourceList, ObjectSpace, Sub(o)
				o.OwnerKey = Me.Oid
			End Sub)
			_MainFeature = If(MainFeatureName Is Nothing, Nothing, Features.FirstOrDefault(Function(f) f.Name = MainFeatureName))
		End Sub
		Protected Overrides Sub OnSaving()
			FeatureList = SerializationHelper.Save(Features)
			ResourceList = SerializationHelper.Save(Resources)
			MyBase.OnSaving()
		End Sub
		Private _objectSpace As IObjectSpace
		Protected ReadOnly Property ObjectSpace() As IObjectSpace
			Get
				Return _objectSpace
			End Get
		End Property
		Private Property IObjectSpaceLink_ObjectSpace() As IObjectSpace Implements IObjectSpaceLink.ObjectSpace
			Get
				Return _objectSpace
			End Get
			Set(ByVal value As IObjectSpace)
				If _objectSpace IsNot value Then
					_objectSpace = value
				End If
			End Set
		End Property
	End Class
End Namespace
